### PR TITLE
Alternate Solution (https://usaco.guide/silver/func-graphs?lang=java#example---badge)

### PR DESCRIPTION
**Pull request by abhideepd**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._
~~Updating the changes, as per the comments in this pr: #4335~~

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [x] I have linked this PR to any issues that it closes. #4330

### CODE REVIEW

File: content/3_Silver/Func_Graphs.mdx [hunk 1]
```diff
@@ -716,6 +716,71 @@ print(" ".join(str(i + 1) for i in ans))
 </PySection>
 </LanguageSection>
 
+### Solution 3
+
+It's also possible to use `floyd(x)` to generate answers for all vertices in the connected component
+containing `x` without using adjacency lists.
```
Comment (SansPapyrus683) [inline content/3_Silver/Func_Graphs.mdx:722]: what do you use instead of adjacency lists then? explain here

Comment (SansPapyrus683) [reply content/3_Silver/Func_Graphs.mdx:722]: wait hold on i read your other comment- what's the complexity of this solution?

File: content/3_Silver/Func_Graphs.mdx [hunk 2]
```diff
@@ -716,6 +716,72 @@ print(" ".join(str(i + 1) for i in ans))
...
+	static void floyd(int curr) {
+		int a = arr[curr];
+		int b = arr[arr[curr]];
+		while (a != b) {
+			a = arr[a];
+			b = arr[arr[b]];
+		}
+		a = curr;
+		// finds the entry cycle node and fills the ans array of nodes in the
+		// cycle
```
Comment (ryanchou-dev) [inline content/3_Silver/Func_Graphs.mdx:765]: multiline like:
```
/*
 * hi
 * hola
 */
```

**Review by ryanchou-dev — COMMENTED** [review]
im not sure if this is different enough of a solution to justify another section. maybe an [info section](https://usaco.guide/general/working-mdx#info-block) would be better

bump on kevin's comments too!

File: content/3_Silver/Func_Graphs.mdx (hunk 2 above)
Comment (abhideepd) [reply content/3_Silver/Func_Graphs.mdx]: updated

File: content/3_Silver/Func_Graphs.mdx (hunk 1 above)
Comment (abhideepd) [reply content/3_Silver/Func_Graphs.mdx]: It seems, I have caused a bit of chaos and confusion. Please ignore the previous pr, which I had earlier mentioned.

My approach:
So, I am using floyd, to find the cycles and then update the coordinates of non cycles and cycle vertices. While, finding a cycle from a, if we encounter ans[b] to not be -1, this means, this coordinate has been already calculated. In this case, the fillUp() method is called. There are two scenarios, either this coordinate (b) is in a cycle, or a non cycle. Whatever might be the case, from that node a to this b, the, answer will for all of the nodes will be ans[b].  

I hope, I was clear.

**Comment by abhideepd** [comment]
> bro why'd u close the other pr

Did it by mistake. Please ignore the previous pr. I have updated this pr, with the relevant changes.

**Comment by abhideepd** [comment]
> im not sure if this is different enough of a solution to justify another section. maybe an [info section](https://usaco.guide/general/working-mdx#info-block) would be better
> 
> bump on kevin's comments too!

This approach is a bit different, kind of a variation. I feel, including multiple approaches and variations to a solution, helps the reader, achieve more clarity of the concepts.  Also, this area is collapsible, so it wouldn't hamper the ui aesthetic of the website 😬

File: content/3_Silver/Func_Graphs.mdx [hunk 3]
```diff
@@ -715,73 +715,98 @@ print(" ".join(str(i + 1) for i in ans))
 
 </PySection>
 </LanguageSection>
-
 ### Solution 3
-
-It's also possible to use `floyd(x)` to generate answers for all vertices in the connected component
-containing `x` without using adjacency lists.
-
+Without using any adjacency lists. `floyd(x)` generates answers for all connected vertices. `fillup(a,b)` fills up the vertices, from `a` to `b`, `b` being the coordinate, whose `ans[b]` has already been calculated `floyd(x)`
```
Comment (SansPapyrus683) [inline content/3_Silver/Func_Graphs.mdx]: ayo why'd you delete all my stuff

Comment (abhideepd) [reply content/3_Silver/Func_Graphs.mdx]: Updated your changes. Sorry, had overlooked your commit, above.

Comment (abhideepd) [reply content/3_Silver/Func_Graphs.mdx]: Is there some precommit hook, that checks, for the length of the text on each row ? It seems to be updating wrt to that. Checked it multiple times with multiple commits, in this pr. Attaching screenshot.

image

File: content/3_Silver/Func_Graphs.mdx [hunk 4]
```diff
@@ -715,7 +715,105 @@ print(" ".join(str(i + 1) for i in ans))
 
 </PySection>
 </LanguageSection>
+### Solution 3
 
+Without using any adjacency lists `floyd(x)` generates answers for all connected vertices. `fillup(a,b)` fills up the vertices, from `a` to `b`, `b` being the coordinate, whose `ans[b]` has already been calculated `floyd(x)`
```
Comment (SansPapyrus683) [inline content/3_Silver/Func_Graphs.mdx:720]: i still don't see my changes in this

Comment (abhideepd) [reply content/3_Silver/Func_Graphs.mdx:720]: implemented the changes, looking through your previous commit [here](https://github.com/cpinitiative/usaco-guide/pull/4336#:~:text=the%20other%20pr-,Update%20Func_Graphs.mdx,-Verified)

Also, updated this branch, which was 2 commits behind [here](https://github.com/cpinitiative/usaco-guide/commits/master/#:~:text=Merge%20pull%20request%20%234370%20from%20cpinitiative/dependabot/npm_and_yarn/jose%2D2.0.7)

am i missing something ?

File: content/3_Silver/Func_Graphs.mdx (hunk 1 above)
Comment (SansPapyrus683) [reply content/3_Silver/Func_Graphs.mdx:722]: > however, its not always trusted...

wdym by this

Comment (abhideepd) [reply content/3_Silver/Func_Graphs.mdx:722]: i meant, the runtime metric no. shown by the online judges in general is not trustable (specially leetcode). have been advised by people around, and blogs in general. personally, i have found a lot of time discrepancy on leetcode, but not so much on codeforces. for this problem, i ran it twice today, and it gave a 30ms difference. not that much, but still. so, i don't focus too much on that runtime metric number.

File: content/3_Silver/Func_Graphs.mdx [hunk 5]
```diff
@@ -716,6 +716,97 @@ print(" ".join(str(i + 1) for i in ans))
 </PySection>
 </LanguageSection>
 
+### Solution 3
+
+It's also possible to use `floyd(x)` to generate answers for all vertices in the connected component
+containing `x` without using adjacency lists.
+
+Without using any adjacency lists `floyd(x)` generates answers for all connected vertices. `fillup(a,b)` fills up the vertices, from `a` to `b`, `b` being the coordinate, whose `ans[b]` has already been calculated `floyd(x)`
```
Comment (SansPapyrus683) [inline content/3_Silver/Func_Graphs.mdx:724]:
```suggestion
```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
@ryanchou-dev review?

File: content/3_Silver/Func_Graphs.mdx [hunk 6]
```diff
@@ -716,6 +716,95 @@ print(" ".join(str(i + 1) for i in ans))
 </PySection>
 </LanguageSection>
 
+### Solution 3
```
Comment (ryanchou-dev) [inline content/3_Silver/Func_Graphs.mdx:719]:
```suggestion
```

File: content/3_Silver/Func_Graphs.mdx [hunk 7]
```diff
@@ -716,6 +716,95 @@ print(" ".join(str(i + 1) for i in ans))
...
+		for (int i = 1; i <= n; i++) { output.append(ans[i]).append(" "); }
+		output.setLength(output.length() - 1);
+		System.out.println(output);
+	}
+
+	/**
+	 * fills up all vertices from curr to firstCycle,
+	 * whose answer has already been calculated
+	 */
+	static void fillUp(int curr, int firstCycle) {
```
Comment (ryanchou-dev) [inline content/3_Silver/Func_Graphs.mdx:760]: `snake_case` for c++ -- rename to `fill_up`?

Comment (ryanchou-dev) [reply content/3_Silver/Func_Graphs.mdx:760]: same for another variable names too ^

**Comment by ryanchou-dev** [comment]
> > im not sure if this is different enough of a solution to justify another section. maybe an [info section](https://usaco.guide/general/working-mdx#info-block) would be better
> > bump on kevin's comments too!
> 
> This approach is a bit different, kind of a variation. I feel, including multiple approaches and variations to a solution, helps the reader, achieve more clarity of the concepts. Also, this area is collapsible, so it wouldn't hamper the ui aesthetic of the website 😬

well, conceptually it's the same right? using floyd to find cycles + filling up the other nodes.

i still think it'd be better as:

```
<solution 2 code>

It's also possible to use floyd(x) to generate answers for all vertices in the connected component containing x without using adjacency lists. We can do this by ...

<psuedo code of fill_up function>
```

if @SansPapyrus683 thinks it's good then that's okay too

File: content/3_Silver/Func_Graphs.mdx [hunk 8]
```diff
@@ -716,6 +716,95 @@ print(" ".join(str(i + 1) for i in ans))
...
+		for (int i = 1; i <= n; i++) { output.append(ans[i]).append(" "); }
+		output.setLength(output.length() - 1);
+		System.out.println(output);
+	}
+
+	/**
+	 * fills up all vertices from curr to firstCycle,
+	 * whose answer has already been calculated
+	 */
+	static void fill_up(int curr, int first_cycle) {
```
Comment (SansPapyrus683) [inline content/3_Silver/Func_Graphs.mdx:758]: use camelCase

Comment (abhideepd) [reply content/3_Silver/Func_Graphs.mdx:758]: previously used camelCase, @ryanchou-dev  suggested to use snake case [above](https://github.com/cpinitiative/usaco-guide/pull/4336#discussion_r1518676540)

Comment (SansPapyrus683) [reply content/3_Silver/Func_Graphs.mdx:758]: oof uh very sorry about that, could you change it back

File: content/3_Silver/Func_Graphs.mdx (hunk 7 above)
Comment (SansPapyrus683) [reply content/3_Silver/Func_Graphs.mdx]: ryan this is java :skull:

File: content/3_Silver/Func_Graphs.mdx (hunk 8 above)
Comment (abhideepd) [reply content/3_Silver/Func_Graphs.mdx]: updated

**Review by SansPapyrus683 — APPROVED** [review]
acceptable

**Review by ryanchou-dev — APPROVED** [review]
ok 👍

**Pull request merged**